File: app/src/main/java/com/stevenfrew/beatprompter/ImageLine.java
package com.stevenfrew.beatprompter;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.Typeface;

import java.util.ArrayList;
import java.util.Collection;

class ImageLine extends Line
{
    private ImageFile mImageFile;
    private ImageScalingMode mScalingMode;
    private Rect mSourceRect,mDestRect;
    private Bitmap mBitmap;

    ImageLine(ImageFile image, ImageScalingMode scalingMode, Collection<Tag> lineTags, int bars, ColorEvent lastColor, int bpb, int scrollbeat, int scrollbeatOffset, ScrollingMode scrollingMode, ArrayList<FileParseError> parseErrors) {
        super(lineTags, bars, lastColor, bpb, scrollbeat, scrollbeatOffset, scrollingMode, parseErrors);
        mImageFile=image;
        mScalingMode=scalingMode;
    }

    LineMeasurements doMeasurements(Paint paint, float minimumFontSize, float maximumFontSize, int screenWidth, int screenHeight, Typeface font, int highlightColour, int defaultHighlightColour, ArrayList<FileParseError> errors, ScrollingMode scrollMode,CancelEvent cancelEvent)
    {
        String path=mImageFile.mFile.getAbsolutePath();
        BitmapFactory.Options options = new BitmapFactory.Options();
        try {
            mBitmap=BitmapFactory.decodeFile(path, options);
        }
        catch(Exception e) {
            errors.add(new FileParseError(null,SongList.getContext().getString(R.string.could_not_read_image_file)+": "+mImageFile.mName));
            return null;
        }

        int imageHeight=mBitmap.getHeight();
        int imageWidth=mBitmap.getWidth();
        int scaledImageHeight=imageHeight;
        int scaledImageWidth=imageWidth;

        if ((imageWidth > screenWidth)||(mScalingMode==ImageScalingMode.Stretch))
        {
            scaledImageHeight = (int) (imageHeight * ((double) screenWidth / (double) imageWidth));
            scaledImageWidth = screenWidth;
        }

        ArrayList<Integer> graphicHeights = new ArrayList<>();
        graphicHeights.add(scaledImageHeight);
        mSourceRect=new Rect(0,0,imageWidth,imageHeight);
        mDestRect=new Rect(0,0,scaledImageWidth,scaledImageHeight);
        return new LineMeasurements(1, mDestRect.width(), mDestRect.height(), graphicHeights, highlightColour, mLineEvent, mNextLine, mYStartScrollTime,scrollMode);
    }

    Collection<LineGraphic> getGraphics(boolean allocate)
    {
        for(int f=0;f<mLineMeasurements.mLines;++f)
        {
            LineGraphic graphic=mGraphics.get(f);
            if ((graphic.mLastDrawnLine != this) &&(allocate))
            {
                Paint paint = new Paint();
                Canvas c = new Canvas(graphic.mBitmap);
                c.drawBitmap(mBitmap,mSourceRect,mDestRect,paint);
                graphic.mLastDrawnLine = this;
            }
        }
        return mGraphics;
    }

    void recycleGraphics()
    {
        super.recycleGraphics();
        mBitmap.recycle();
    }
}

File: app/src/main/kotlin/com/stevenfrew/beatprompter/ImageLine.kt
package com.stevenfrew.beatprompter

import android.graphics.*
import com.stevenfrew.beatprompter.cache.ImageFile

class ImageLine internal constructor(mImageFile:ImageFile, scalingMode:ImageScalingMode, lineTime:Long, lineDuration:Long, beatInfo:LineBeatInfo, displaySettings:SongDisplaySettings, pixelPosition:Int, startScrollTime:Long, stopScrollTime:Long) : Line(lineTime,lineDuration,beatInfo,pixelPosition,startScrollTime,stopScrollTime) {
    private val mBitmap:Bitmap=BitmapFactory.decodeFile(mImageFile.mFile.absolutePath, BitmapFactory.Options())
    private val mSourceRect: Rect = Rect(0,0,mImageFile.mWidth,mImageFile.mHeight)
    private val mDestinationRect=getDestinationRect(mBitmap,displaySettings.mScreenSize,scalingMode)
    override val mMeasurements:LineMeasurements

    init
    {
        mMeasurements=LineMeasurements(1, mDestinationRect.width(), mDestinationRect.height(), intArrayOf(mDestinationRect.height()), lineTime,lineDuration, startScrollTime, beatInfo.mScrollMode)
    }

    override fun renderGraphics()  {
        for (f in 0 until mMeasurements.mLines) {
            val graphic = mGraphics[f]
            if (graphic.mLastDrawnLine !== this) {
                val paint = Paint()
                val c = Canvas(graphic.mBitmap)
                c.drawBitmap(mBitmap, mSourceRect, mDestinationRect, paint)
                graphic.mLastDrawnLine = this
            }
        }
    }

    override fun recycleGraphics() {
        mBitmap.recycle()
        super.recycleGraphics()
    }

    companion object {
        private fun getDestinationRect(bitmap:Bitmap,screenSize:Rect,scalingMode:ImageScalingMode):Rect
        {
            val imageHeight = bitmap.height
            val imageWidth = bitmap.width
            var scaledImageHeight = imageHeight
            var scaledImageWidth = imageWidth

            if (imageWidth > screenSize.width() || scalingMode === ImageScalingMode.Stretch) {
                scaledImageHeight = (imageHeight * (screenSize.width().toDouble() / imageWidth.toDouble())).toInt()
                scaledImageWidth = screenSize.width()
            }

            return Rect(0, 0, scaledImageWidth, scaledImageHeight)
        }
    }
}